Adding nagios library

hyperbolic2346 opened the pull request:
Specifically, functions to install and remove plugins.

PRing this here instead of upstream due to issue in launchpad where I am unable to make a merge proposal. URL https://code.launchpad.net/~knobby/nagios-layer/+git/nagios-layer/+ref/mwilson/utils appears broken currently.

hyperbolic2346 committed:
Adding nagios library

Specifically, functions to install and remove plugins.

tvansteenburgh commented on lib/charms/layer/nagios.py:
```diff
@@ -0,0 +1,53 @@
+    Args:
+        source_path: Source file path to install
+        plugin_name: Name of the plugin in nagios
+
```
Please include `data` in the Args section - it's not clear what it is. It overrides the contents of `source_path`?

tvansteenburgh commented on lib/charms/layer/nagios.py:
```diff
@@ -0,0 +1,53 @@
+        if dest == data:
+            # same file
+            return dest_path
+        # else:
```
Are you intending to handle the 'else' case?

hyperbolic2346 replied:
Originally I had an else here to complain about duplicate plugin names, but then I realized through my testing that it will be common to rewrite the plugin as configuration changes. This results in being unable to tell the difference, so I removed the error.

hyperbolic2346 committed:
Updates based on review comments

tvansteenburgh commented on lib/charms/layer/nagios.py:
```diff
@@ -0,0 +1,57 @@
+NAGIOS_PLUGINS_DIR = '/usr/lib/nagios/plugins'
+
+
+def install_nagios_plugin(source_path, plugin_name, data=None):
```
This API doesn't make sense. `source_path` is a required arg, but it's not used if `data` is passed.

hyperbolic2346 replied:
This began life as `def install_nagios_plugin(source_path, plugin_name)` and then I realized I needed to template my plugin, which made me then think about writing it to disk and then reading it right back to copy it and that didn't make sense to me. I added data as a way around that. I thought about making source_path either the data or a filename, but I didn't want to rely on filename parsing to guess. Do you have any suggestions here?

johnsca replied:
I think I lean toward:

```python
def install_nagios_plugin(data, plugin_name):
```

I don't think it's too onerous to require the end-user to do:

```python
install_nagios_plugin(Path(source_file).read_text(), plugin_name)
```

tvansteenburgh replied:
You could also have:
```python
def install_nagios_plugin_from_text(text, plugin_name):
    # rest of code

def install_nagios_plugin_from_file(source_file_path, plugin_name):
    install_nagios_plugin_from_text(Path(source_file_path).read_text(), plugin_name)
```

hyperbolic2346 committed:
Updated after review

Cleaned up API to indicate reading from file or using text
Converted to pathlib

tvansteenburgh commented on lib/charms/layer/nagios.py:
```diff
@@ -0,0 +1,60 @@
+    """ Install a nagios plugin.
+
+    Args:
+        test: Source file text to install
```
Suggestion:
```
text: Plugin source code (str)
```

tvansteenburgh commented on lib/charms/layer/nagios.py:
```diff
@@ -0,0 +1,60 @@
+    """ Install a nagios plugin.
+
+    Args:
+        source_file_path: File that is copied to create plugin
```
Suggestion:
```
source_file_path: Path to plugin source file
```

tvansteenburgh commented on lib/charms/layer/nagios.py:
```diff
@@ -0,0 +1,60 @@
+    """ Remove a nagios plugin.
+
+    Args:
+        plugin_name: Plugin name that was passed to install_plugin
```
Suggestion:
```
plugin_name: Name of the plugin in nagios
```
(This matches the other doc strings.)

tvansteenburgh reviewed (commented):
Some minor suggestions but otherwise LGTM, thanks for the revisions.

hyperbolic2346 committed:
Updates to doc strings per review
